README.md
# Adaptive Modular Scale

[![npm version](https://badge.fury.io/js/adaptive-modular-scale.svg)](https://www.npmjs.com/package/adaptive-modular-scale)
[![Codecov Coverage](https://img.shields.io/codecov/c/github/jeslage/adaptive-modular-scale.svg?style=shield)](https://codecov.io/gh/jeslage/adaptive-modular-scale/)
[![CircleCI](https://circleci.com/gh/jeslage/adaptive-modular-scale.svg?style=shield)](https://circleci.com/gh/jeslage/adaptive-modular-scale)

A small javascript helper to get an adaptive modular scale font-size css string. It will interpolate between two modular scales and breakpoints and will return you sizes based on the given modular scales. This package is inspired by this [tool](https://codepen.io/tol-is/pen/mQVLPY) by Florian Schulz.

```bash
yarn add adaptive-modular-scale

npm install adaptive-modular-scale
```

## Usage

```js
import styled from 'styled-components';
import adaptiveModularScale from 'adaptive-modular-scale';

const Headline = styled.h1`
  ${adaptiveModularScale(3, {
    // Minimum and maximum base
    base: [14, 16],
    // Minimum and maximum ratio
    ratio: [1.2, 1.78],
    // Two screen widths to interpolate between
    width: [320, 960],
    // Number of breakpoints between two widths
    breakpoints?: 8,
    // Optional corrections for steps
    corrections?: {
      // Correct minimal and maximum size for step. The below example
      // will correct the minimal step by -1px and the maximum step by +2px
      3: [-1, 2]
    },
    lineHeights?: {
      // Add minimal and maximum lineHeights for step.
      3: [1.2, 1.5]
    }
  })}
`;
```

The above example will return the following css.

```css
.sc-bZQynM {
  font-size: 1.45rem;
  line-height: 1.2;

  @media (min-width: 25rem) {
    font-size: 1.989rem;
    line-height: 1.2375;
  }
  @media (min-width: 30rem) {
    font-size: 2.528rem;
    line-height: 1.275;
  }
  @media (min-width: 35rem) {
    font-size: 3.068rem;
    line-height: 1.3125;
  }
  @media (min-width: 40rem) {
    font-size: 3.607rem;
    line-height: 1.35;
  }
  @media (min-width: 45rem) {
    font-size: 4.147rem;
    line-height: 1.3875;
  }
  @media (min-width: 50rem) {
    font-size: 4.686rem;
    line-height: 1.425;
  }
  @media (min-width: 55rem) {
    font-size: 5.225rem;
    line-height: 1.4625;
  }
  @media (min-width: 60rem) {
    font-size: 5.765rem;
    line-height: 1.5;
  }
}
```

## ThemeProvider

You can also add an `adaptiveModularScale` key to the theme object of your ThemeProvider.

**Example**

```js
import styled, { ThemeProvider } from 'styled-components';
import adaptiveModularScale from 'adaptive-modular-scale';

const theme = {
  adaptiveModularScale: {
    base: [14, 16],
    ratio: [1.2, 1.78],
    width: [320, 960],
    breakpoints: 8
  }
};

// If you added the modular scale values to the ThemeProvider
// you dont have to pass it down every time
const Headline = styled.h1`
  ${adaptiveModularScale(5)}
`;

const App = () => (
  <ThemeProvider theme={theme}>
    <Headline>An adaptive modular scale sized headline</Headline>
  </ThemeProvider>
);
```

## Custom css property

By default `adaptive-modular-scale` uses `font-size` as the css property. But you can also change the property by adding a property key to your config object.

**Example**

```js
import styled, { ThemeProvider } from 'styled-components';
import adaptiveModularScale from 'adaptive-modular-scale';

const Headline = styled.h1`
  ${adaptiveModularScale(5, { property: 'margin-bottom' })}
`;
```

## Options

### `step`

`number` | required

The step on the modular scale

### `config`

`object` | required

### `config.base`

`[number, number]` | required

Array of minimum and maximum base size of modular scale.

### `config.ratio`

`[number, number]` | required

Array of minimum and maximum ratio of modular scale.

### `config.width`

`[number, number]` | required

Array of two screen widths in px. Between these values the modular scale will interpolate between both scales based on screen width.

### `config.breakpoints`

`number` | required

Number of breakpoints between the two given screen widths which will be rendered to the css string.

### `config.corrections`

`{ [step]: [number, number] }` | optional

Add optional corrections array with steps as keys.

```js
const corrections = {
  // This will correct the minimal size of step "0" by -2px
  // and the maximum size by +1px
  0: [-2, 1]
};
```

### `config.lineHeights`

`{ [step]: [number, number] }` | optional

Add optional lineHeights array with steps as keys.

### `config.property`

`string` | optional | default: "font-size"

Add optional css property if you want to use adaptive modular scale for another property than font-size.

## Additional methods

`adaptive-modular-scale` also exports a default modular scale method.

```js
import styled from 'styled-components';
import { modularScale } from 'adaptive-modular-scale';

const base = 16;
const ratio = 1.6;

const Headline = styled.h1`
  font-size: ${modularScale(5, base, ratio)}px;
`;
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.2.0] - 2020-02-22

### Added

- `lineHeights` object to config
- Jest tests and build ci pipeline

## [1.1.1] - 2020-01-22

### Fixed

- `property` config key

## [1.1.0] - 2020-01-22

### Added

- `corrections` object to config
- `property` key to config

## [1.0.1] - 2020-01-19

### Updated

- Moved to math.pow for modular scale calculation due to wrong calculations
- Rounded resulting values with toFixed(3)

## [1.0.0-rc.1] - 2020-01-18

### Fixed

- Wrong repository and issue key in package.json

## 1.0.0 - 2020-01-18

- Initial release

[unreleased]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.2.0...develop
[1.2.0]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.1.1...v1.2.0
[1.1.1]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.0.1...v1.1.0
[1.0.1]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.0.0-rc.1...v1.0.1
[1.0.0-rc.1]: https://github.com/jeslage/adaptive-modular-scale/compare/v1.0.0...v1.0.0-rc.1


src/__tests__/index.spec.ts
import 'jest';
import { adaptiveModularScale } from '../index';

test('should return string', () => {
  expect(
    adaptiveModularScale(3, {
      ratio: [1.2, 1.67],
      base: [16, 16],
      width: [375, 1440],
      breakpoints: 10,
      lineHeights: {
        '3': [1, 1.5]
      }
    })()
  ).toBe(
    'font-size: 1.728rem; line-height: 1; @media (min-width: 30.094rem) { font-size: 2.021rem; line-height: 1.05; }@media (min-width: 36.750rem) { font-size: 2.314rem; line-height: 1.1; }@media (min-width: 43.406rem) { font-size: 2.607rem; line-height: 1.15; }@media (min-width: 50.063rem) { font-size: 2.900rem; line-height: 1.2; }@media (min-width: 56.719rem) { font-size: 3.193rem; line-height: 1.25; }@media (min-width: 63.375rem) { font-size: 3.486rem; line-height: 1.3; }@media (min-width: 70.031rem) { font-size: 3.779rem; line-height: 1.35; }@media (min-width: 76.688rem) { font-size: 4.072rem; line-height: 1.4; }@media (min-width: 83.344rem) { font-size: 4.365rem; line-height: 1.45; }@media (min-width: 90.000rem) { font-size: 4.657rem; line-height: 1.5; }'
  );
});

test('should use theme', () => {
  expect(
    adaptiveModularScale(3)({
      theme: {
        adaptiveModularScale: {
          ratio: [1.2, 1.67],
          base: [16, 16],
          width: [375, 1440],
          breakpoints: 10,
          lineHeights: {
            '3': [1, 1.5]
          }
        }
      }
    })
  ).toBe(
    'font-size: 1.728rem; line-height: 1; @media (min-width: 30.094rem) { font-size: 2.021rem; line-height: 1.05; }@media (min-width: 36.750rem) { font-size: 2.314rem; line-height: 1.1; }@media (min-width: 43.406rem) { font-size: 2.607rem; line-height: 1.15; }@media (min-width: 50.063rem) { font-size: 2.900rem; line-height: 1.2; }@media (min-width: 56.719rem) { font-size: 3.193rem; line-height: 1.25; }@media (min-width: 63.375rem) { font-size: 3.486rem; line-height: 1.3; }@media (min-width: 70.031rem) { font-size: 3.779rem; line-height: 1.35; }@media (min-width: 76.688rem) { font-size: 4.072rem; line-height: 1.4; }@media (min-width: 83.344rem) { font-size: 4.365rem; line-height: 1.45; }@media (min-width: 90.000rem) { font-size: 4.657rem; line-height: 1.5; }'
  );
});

test('should use custom css property', () => {
  expect(
    adaptiveModularScale(3, {
      ratio: [1.2, 1.67],
      base: [16, 16],
      width: [375, 1440],
      breakpoints: 10,
      property: 'margin-top'
    })()
  ).toBe(
    'margin-top: 1.728rem;  @media (min-width: 30.094rem) { margin-top: 2.021rem;  }@media (min-width: 36.750rem) { margin-top: 2.314rem;  }@media (min-width: 43.406rem) { margin-top: 2.607rem;  }@media (min-width: 50.063rem) { margin-top: 2.900rem;  }@media (min-width: 56.719rem) { margin-top: 3.193rem;  }@media (min-width: 63.375rem) { margin-top: 3.486rem;  }@media (min-width: 70.031rem) { margin-top: 3.779rem;  }@media (min-width: 76.688rem) { margin-top: 4.072rem;  }@media (min-width: 83.344rem) { margin-top: 4.365rem;  }@media (min-width: 90.000rem) { margin-top: 4.657rem;  }'
  );
});

test('should use default breakpoints', () => {
  expect(
    adaptiveModularScale(3, {
      ratio: [1.2, 1.67],
      base: [16, 16],
      width: [375, 1440]
    })()
  ).toBe(
    'font-size: 1.728rem;  @media (min-width: 31.758rem) { font-size: 2.094rem;  }@media (min-width: 40.078rem) { font-size: 2.460rem;  }@media (min-width: 48.398rem) { font-size: 2.827rem;  }@media (min-width: 56.719rem) { font-size: 3.193rem;  }@media (min-width: 65.039rem) { font-size: 3.559rem;  }@media (min-width: 73.359rem) { font-size: 3.925rem;  }@media (min-width: 81.680rem) { font-size: 4.291rem;  }@media (min-width: 90.000rem) { font-size: 4.657rem;  }'
  );
});

test('should use corrections object', () => {
  expect(
    adaptiveModularScale(3, {
      ratio: [1.2, 1.67],
      base: [16, 16],
      width: [375, 1440],
      corrections: {
        '3': [1, -2]
      }
    })()
  ).toBe(
    'font-size: 1.790rem;  @media (min-width: 31.758rem) { font-size: 2.133rem;  }@media (min-width: 40.078rem) { font-size: 2.476rem;  }@media (min-width: 48.398rem) { font-size: 2.819rem;  }@media (min-width: 56.719rem) { font-size: 3.161rem;  }@media (min-width: 65.039rem) { font-size: 3.504rem;  }@media (min-width: 73.359rem) { font-size: 3.847rem;  }@media (min-width: 81.680rem) { font-size: 4.190rem;  }@media (min-width: 90.000rem) { font-size: 4.532rem;  }'
  );
});

test('should throw error', () => {
  expect(adaptiveModularScale(0)).toThrowError(
    'No config or theme object with adaptiveModularScale key provided.'
  );
});


src/helper.ts
export const calculateSize = (
  width: number,
  from: number,
  to: number,
  min: number,
  max: number
): number => {
  return Math.round(min + (max - min) * ((width - from) / (to - from)));
};

/**
 * Returns modular scale value
 * @param  {number} steps - Step of the modular scale
 * @param  {number} base - Base size of the modular scale
 * @param  {number} ratio - Ratio of the modular scale
 * @example modularScale(3, 16, 1.67)
 * @returns {number} Resulting modular scale value
 */
export const modularScale = (
  steps: number,
  base: number,
  ratio: number
): number => Math.round(base * ratio ** steps);

/**
 * Transform px value to rem value.
 * @param {number} px - PX value as a integer. e.g. 12
 * @returns {string} - PX value in REM
 * @example ${px2rem(25)}
 */
export const px2rem = (px: number): string => {
  return `${px / 16}rem`;
};


src/index.ts
export interface Theme {
  adaptiveModularScale: Config;
  [key: string]: any;
}

export interface FunctionProps {
  theme: Theme;
  [key: string]: any;
}

export interface Config {
  base: [number, number];
  ratio: [number, number];
  width: [number, number];
  breakpoints?: number;
  corrections?: {
    [key: string]: [number, number];
  };
  lineHeights?: {
    [key: string]: [number, number];
  };
  property?: string;
}

const calcSize = (
  width: number,
  from: number,
  to: number,
  min: number,
  max: number
): number => {
  return min + (max - min) * ((width - from) / (to - from));
};

/**
 * Returns modular scale value
 * @param  {number} steps - Step of the modular scale
 * @param  {number} base - Base size of the modular scale
 * @param  {number} ratio - Ratio of the modular scale
 * @example modularScale(3, 16, 1.67)
 * @returns {number} Resulting modular scale value
 */
export const modularScale = (steps: number, base: number, ratio: number): number =>
  base * Math.pow(ratio, steps);

/**
 * Transform px value to rem value.
 * @param {number} px - PX value as a integer. e.g. 12
 * @returns {string} - PX value in REM
 * @example ${px2rem(25)}
 */
export const px2rem = (px: number): string => `${(px / 16).toFixed(3)}rem`;

/**
 * Returns adaptive modular scale css font-size string
 * @param  {number} step - Step of the modular scale
 * @param  {object} config - Config object with base, ratio, width and breakpoints key
 * @example adaptiveModularScale(3, { base: [14, 16], ratio: [1.4, 1.67], width: [320, 960], breakpoints: 10 })
 * @returns {string} Resulting adaptive modular scale css font-size string
 */
export const adaptiveModularScale = (step: number, config?: Config) => (
  props?: FunctionProps
): string => {
  const settings = props?.theme?.adaptiveModularScale || config;

  if (typeof settings === 'undefined') {
    throw Error('No config or theme object with adaptiveModularScale key provided.');
  }

  const { width, base, ratio, breakpoints, corrections, lineHeights } = settings;

  const cssProp = config && config.property ? config.property : 'font-size';

  const lineHeight = lineHeights && lineHeights[step] ? lineHeights[step] : undefined;
  const correction = corrections && corrections[step] ? corrections[step] : [0, 0];

  const minSize = modularScale(step, base[0], ratio[0]) + correction[0];
  const maxSize = modularScale(step, base[1], ratio[1]) + correction[1];

  const steps = (width[1] - width[0]) / (breakpoints || 8);

  let mediaQueries = ``;

  for (let i = width[0] + steps; i <= width[1]; i += steps) {
    const size = calcSize(i, width[0], width[1], minSize, maxSize);

    const lh = lineHeight
      ? `line-height: ${calcSize(i, width[0], width[1], lineHeight[0], lineHeight[1])};`
      : '';

    const mq = px2rem(i);
    const fontSize = px2rem(size);

    mediaQueries += `@media (min-width: ${mq}) { ${cssProp}: ${fontSize}; ${lh} }`;
  }

  return `${cssProp}: ${px2rem(minSize)}; ${
    lineHeight ? `line-height: ${lineHeight[0]};` : ''
  } ${mediaQueries}`;
};

export default adaptiveModularScale;
